### PR TITLE
implement http 400 bad request

### LINKED ISSUE #1085: api/http: send HTTP code 400 on errors

### PR DESCRIPTION
**Pull request by howjmay**
Add arg `status_code` to `error_serialize_response()`
Return status code 400 if no message and no status code was specified.

fixes #1085

### CODE REVIEW

**Comment by thibault-martinez** [comment]
Sorry for the delay, I had to think about this one. Will review later today !

File: ciri/api/http/http.c [hunk 1]
```diff
@@ -28,14 +28,17 @@ typedef struct iota_api_http_session_s {
...
 
 static retcode_t error_serialize_response(iota_api_http_t *const http, error_res_t **const error,
-                                          char const *const message, char_buffer_t *const out) {
+                                          char const *const message, uint16_t const status_code,
+                                          char_buffer_t *const out) {
   retcode_t ret = RC_OK;
 
   if (*error == NULL) {
-    if (message == NULL) {
+    if (message == NULL && status_code == MHD_HTTP_INTERNAL_SERVER_ERROR) {
```
Comment (thibault-martinez) [inline ciri/api/http/http.c]: Why this `&&` ?

File: ciri/api/http/http.c [hunk 2]
```diff
@@ -28,14 +28,17 @@ typedef struct iota_api_http_session_s {
...
       *error = error_res_new("Internal server error");
```
Comment (thibault-martinez) [inline ciri/api/http/http.c:37]: Can you add `status_code` as parameter to `error_res_new` ?

File: ciri/api/http/http.c [hunk 3]
```diff
@@ -586,12 +590,13 @@ static int iota_api_http_handler(void *cls, struct MHD_Connection *connection, c
...
 
-    iota_api_http_process_request(api, json_item->valuestring, sess->request->data, response_buf);
+    ret = iota_api_http_process_request(api, json_item->valuestring, sess->request->data, response_buf);
   }
 
   response = MHD_create_response_from_buffer(response_buf->length, response_buf->data, MHD_RESPMEM_MUST_COPY);
+
   MHD_add_response_header(response, MHD_HTTP_HEADER_CONTENT_TYPE, "application/json");
-  ret = MHD_queue_response(connection, MHD_HTTP_OK, response);
+  ret = MHD_queue_response(connection, error->status_code, response);
```
Comment (thibault-martinez) [inline ciri/api/http/http.c]: `MHD_HTTP_OK` if no error, `error->status_code` otherwise

**Review by thibault-martinez — CHANGES_REQUESTED** [review]

File: ciri/api/http/http.c (hunk 1 above)
Comment (howjmay) [reply ciri/api/http/http.c]: I originally thought I should use `Bad request` as dfault error status code.
I will change `&&` to `||`

Comment (thibault-martinez) [reply ciri/api/http/http.c]: I don't think it's necessary at all

Comment (howjmay) [reply ciri/api/http/http.c]: what should I do?

**Comment by thibault-martinez** [comment]
Since you opened this PR, I started working on a gRPC API.
This file will probably change a lot, so please address comments, fix conflicts and I'm not sure when I'll be able to merge it, sorry about that !

File: ciri/api/http/http.c (hunk 1 above)
Comment (howjmay) [reply ciri/api/http/http.c]: remove `message == NULL`？

Comment (thibault-martinez) [reply ciri/api/http/http.c]: Remove `&& status_code == MHD_HTTP_INTERNAL_SERVER_ERROR`

Comment (howjmay) [reply ciri/api/http/http.c]: OK

**Comment by howjmay** [comment]
> Since you opened this PR, I started working on a gRPC API.
> This file will probably change a lot, so please address comments, fix conflicts and I'm not sure when I'll be able to merge it, sorry about that !

got it! cheer up

**Comment by thibault-martinez** [comment]
Won't be merged due to ciri being discontinued and entangled soon being archived.
Thanks anyway!

**Pull request closed without merging**